### DISCUSSION TITLE
I need help with coldfire mcf5372 settings

### DISCUSSION THREAD

**Discussion by kabiri** [NationalSecurityAgency/ghidra#4300] [Q&A]
I have a coldfire mcf5372 binary file.
But Ghidra only has a 68,000 processor in terms of language.
Which processor is suitable for coldfire mcf5372? Also how do I define a coldfire mcf5372 memory map?
Whatever I define, it does not decode
![MemoryMap](https://user-images.githubusercontent.com/8082245/171158667-7c07ac3e-41fe-46e8-90f3-ad2f75217ad4.png)
 correctly.

**Comment by kabiri** [comment]
Has anyone ever done this before?

**Comment by Tim---** [comment]
Could you clarify what you mean by "it doesn't decode" ? 
Are you talking about the instructions that are not decoded, or the memory references that are not created ?

Your memory map looks correct (although the blocks you created are huge !).
A screenshot of your Listing/Decompiled view to pinpoint the problem would probably be helpful.

**Reply by kabiri** [reply]
> 
sorry "it doesn't decode" = The code is not disassembled correctly.

It finds only 3 functions

![image](https://user-images.githubusercontent.com/8082245/176899933-f8e57e9e-a8a5-42f6-8933-924941028d2a.png)

**Comment by Tim---** [comment]
Ok, it makes more sense. 

At first glance, the processor choice looks correct (I don't know the 68000 arch, but `cmp*` followed by a conditional branch looks fine). But I doubt that the functions starts here (beginning with a moveq is either hand-made assembly, or just wrong). So even the functions that the analyzers found are dubious.

Even if the analyzers are lost, you can still disassemble functions by hand and start analyzing your firmware. If you have some processor docs, you can start from the reset vector to see if the code makes sense.

Now, why are the analyzers not finding the functions ? There are a few options:
- The "Function Start Search" analyzer relies on known preambles to find functions. Perhaps the preambles for your processor is different from the preambles for the generic 68000 processor. So you will have to disassemble a number of functions by hand, and find it there is a recurrent preamble that you can use.
- The "Reference"/"Subroutine Reference" analyzers will find functions that are called by other functions. But if your base address is incorrect, this won't work. You loaded the code at address 0, but maybe it is loaded further in memory. You will have to disassemble functions by hand to find the correct base address of your firmware.
- Maybe the ISA implemented in your Coldfire processor is more recent that the one implemented in Ghidra. If you find a sequence of instructions that looks correct, and ends abruptly with an "undefined opcode" error, you can look at the ISA to find if it's a new instruction.
- (The "unreachable block" errors may suggest that the permissions of you "ram" region are incorrect. Maybe this region is read/write ?)

I guess this checklist should cover the basic problems when starting a firmware analysis. It will enable you to narrow down the problem.

Note: also, you may want to add the two overview bars in your Listing view (search for "overview" in the help). This allows you (and us :) ) to have a better understanding of what your code contains, and which parts Ghidra was able to decode.

**Reply by kabiri** [reply]
Thank you for your help and guidance.
I had previously disassemble several functions manually, but the analyzer did not recognize the functions.
I looked for reset and couldn't find it (I'm very new).
I don't know what ISA is and what it does.


> Ok, it makes more sense.
> 
> At first glance, the processor choice looks correct (I don't know the 68000 arch, but `cmp*` followed by a conditional branch looks fine). But I doubt that the functions starts here (beginning with a moveq is either hand-made assembly, or just wrong). So even the functions that the analyzers found are dubious.
> 
> Even if the analyzers are lost, you can still disassemble functions by hand and start analyzing your firmware. If you have some processor docs, you can start from the reset vector to see if the code makes sense.
> 
> Now, why are the analyzers not finding the functions ? There are a few options:
> 
> * The "Function Start Search" analyzer relies on known preambles to find functions. Perhaps the preambles for your processor is different from the preambles for the generic 68000 processor. So you will have to disassemble a number of functions by hand, and find it there is a recurrent preamble that you can use.
> * The "Reference"/"Subroutine Reference" analyzers will find functions that are called by other functions. But if your base address is incorrect, this won't work. You loaded the code at address 0, but maybe it is loaded further in memory. You will have to disassemble functions by hand to find the correct base address of your firmware.
> * Maybe the ISA implemented in your Coldfire processor is more recent that the one implemented in Ghidra. If you find a sequence of instructions that looks correct, and ends abruptly with an "undefined opcode" error, you can look at the ISA to find if it's a new instruction.
> * (The "unreachable block" errors may suggest that the permissions of you "ram" region are incorrect. Maybe this region is read/write ?)
> 
> I guess this checklist should cover the basic problems when starting a firmware analysis. It will enable you to narrow down the problem.
> 
> Note: also, you may want to add the two overview bars in your Listing view (search for "overview" in the help). This allows you (and us :) ) to have a better understanding of what your code contains, and which parts Ghidra was able to decode.
